### PR TITLE
Further work on issue #3018.

### PR DESCRIPTION
**Pull request by julianser**
I implemented proper stack trace copying for a couple of functions, and changed the printing of the stack trace. Now, a stack trace is list of lists of tuples. 

Sorry, I had to create a new pull request. I completely messed up my other branch trying to do rebase myself.

### CODE REVIEW

File: theano/gof/fg.py [hunk 1]
```diff
@@ -334,13 +334,15 @@ def find_path_to(output_var, input_var):
                             assert path is not None
                             tr = getattr(r.tag, 'trace', None)
                             detailed_err_msg = ""
-                            if tr:
-                                sio = StringIO()
-                                traceback.print_list(tr, sio)
-                                tr = sio.getvalue()
+                            if len(tr) > 0:
```
Comment (lamblin) [inline theano/gof/fg.py]: `tr` can be `None` here, so its length is not defined in that case.
The easiest fix would be to change l. 335 to something like `tr = getattr(r.tag, 'trace', [])`.

**Comment by julianser** [comment]
@lamblin It would be nice if we could go through these tomorrow or early next week.

Especially, I'm not sure about local_track_shape_i and local_elemwise_alloc.

**Comment by julianser** [comment]
The test fails because of some white space issues. But other than that everything seems to be OK...

**Comment by lamblin** [comment]
It looks good and tests pass, so I'd like to merge this PR for the moment, then we can look into continuing that work for other optimizations.
Could you rebase on top of the current master to get rid of the "merge" commits in the history?

**Comment by julianser** [comment]
Sounds good. I'll rebase Monday morning, or maybe even tomorrow if I get the time. 

Sorry, I havn't been active this week...

File: theano/tensor/opt.py [hunk 2]
```diff
@@ -519,7 +571,12 @@ def local_lift_transpose_through_dot(node):
     x, y = node.inputs[0].owner.inputs
 
     if x.ndim == y.ndim == 2:
-        return [T.dot(y.T, x.T)]
+        # Output is dot product of transposed inputs in reverse order
+        ret = [T.dot(y.T, x.T)]
+
+        # Copy over stack trace to output from result of dot-product
+        copy_stack_trace(node.inputs[0], ret)
```
Comment (lamblin) [inline theano/tensor/opt.py]: I think we should actually copy the stack trace from node.outputs[0] instead.

**Comment by nouiz** [comment]
travis have errors.

**Comment by lamblin** [comment]
The errors in travis seem to be pep8 formatting, that should be easy to fix.

**Comment by julianser** [comment]
@lamblin Pascal, can you restart the Scrutinizer? It seems to have crashed the other day...

I already have new commits, but I want you to merge this PR before I continue pushing.

- Julian

**Comment by lamblin** [comment]
I don't seem to have the rights to do that yet, but I'll merge anyway. Thanks!

File: theano/gof/fg.py [hunk 3]
```diff
@@ -331,15 +331,17 @@ def find_path_to(output_var, input_var):
...
                                 detailed_err_msg += "\nBacktrace when the variable is created:\n"
-                                detailed_err_msg += str(tr)
 
+                                # Print separate message for each element in
+                                # the list of batcktraces
+                                sio = StringIO()
+                                for subtr in tr:
+                                    traceback.print_list(subtr, sio)
+                                    detailed_err_msg += str(sio.getvalue())
```
Comment (nouiz) [inline theano/gof/fg.py]: sio.getvalue() keep the old content. So this line 344, should be outside the loop.

File: theano/gof/link.py [hunk 4]
```diff
@@ -156,14 +156,16 @@ def raise_with_op(node, thunk=None, exc_info=None, storage_map=None):
...
+    tr = getattr(node.outputs[0].tag, 'trace', [])
+    if len(tr) > 0:
         detailed_err_msg += "\nBacktrace when the node is created:\n"
-        detailed_err_msg += str(tr)
+
+        # Print separate message for each element in the list of batcktraces
+        sio = StringIO()
+        for subtr in tr:
+            traceback.print_list(subtr, sio)
+            detailed_err_msg += str(sio.getvalue())
```
Comment (nouiz) [inline theano/gof/link.py]: idem

File: theano/gof/op.py [hunk 5]
```diff
@@ -458,14 +458,17 @@ def _get_test_value(cls, v):
...
+                if len(tr) > 0:
                     detailed_err_msg += (
                         " \nBacktrace when that variable is created:\n")
-                    detailed_err_msg += str(tr)
+                    # Print separate message for each element in the list 
+                    # of batcktraces
+                    sio = StringIO()
+                    for subtr in tr:
+                        traceback.print_list(subtr, sio)
+                        detailed_err_msg += str(sio.getvalue())
```
Comment (nouiz) [inline theano/gof/op.py]: idem. Should this be refactored to remove the duplicate code?

File: theano/tensor/opt.py [hunk 6]
```diff
@@ -63,6 +63,54 @@
...
+    of tuples. Each tuple contains the filename, line number, function name
+    and so on. Each list of tuples contains the truples belonging to a
+    particular variable.
+    """
+
+    # Store stack traces from from_var
+    tr = []
+    if type(from_var) is list:
+        # If from_var is a list, store concatenated stack traces
+        if len(from_var) > 0:
```
Comment (nouiz) [inline theano/tensor/opt.py]: this if is useless. A loop on an empty list would cause 0 iteration of the loop.

File: theano/tensor/opt.py [hunk 7]
```diff
@@ -63,6 +63,54 @@
...
+            for v in from_var:
+                if hasattr(v.tag, 'trace') and len(v.tag.trace) > 0:
```
Comment (nouiz) [inline theano/tensor/opt.py]: The second condition is not useful. A list + an empty list, return the original list.

File: theano/tensor/opt.py [hunk 8]
```diff
@@ -63,6 +63,54 @@
...
+
+    # Copy over stack traces to to_var
+    if type(to_var) is list:
+        # Copy over stack traces from from_var to each variable in
+        # to_var, including the stack_trace of the to_var before
+        for v in to_var:
+            if hasattr(v.tag, 'trace'):
+                v.tag.trace = v.tag.trace + tr
+            else:
+                v.tag.trace = tr
```
Comment (nouiz) [inline theano/tensor/opt.py]: the 4 lines above can be simplified as:

`v.tag.trace = getattr(v.tag, 'trace', []) + tr`

File: theano/tensor/opt.py (hunk 2 above)
Comment (nouiz) [inline theano/tensor/opt.py]: this comment didn't got answer and wasn't done. Is that normal? At first impression, I think the comment is right.

**Comment by nouiz** [comment]
I didn't checked all change to opt, but I got a few comments.

thanks for this good new feature.

**Pull request merged**